server/container_create.go

```diff
@@ -240,58 +240,93 @@ func buildOCIProcessArgs(containerKubeConfig *pb.ContainerConfig, imageOCIConfig
 
 // setupContainerUser sets the UID, GID and supplemental groups in OCI runtime config
 func setupContainerUser(specgen *generate.Generator, rootfs string, sc *pb.LinuxContainerSecurityContext, imageConfig *v1.Image) error {
-	if sc != nil {
-		containerUser := ""
-		// Case 1: run as user is set by kubelet
-		if sc.GetRunAsUser() != nil {
-			containerUser = strconv.FormatInt(sc.GetRunAsUser().GetValue(), 10)
-		} else {
-			// Case 2: run as username is set by kubelet
-			userName := sc.GetRunAsUsername()
-			if userName != "" {
-				containerUser = userName
-			} else {
-				// Case 3: get user from image config
-				if imageConfig != nil {
-					imageUser := imageConfig.Config.User
-					if imageUser != "" {
-						containerUser = imageUser
-					}
-				}
-			}
-		}
-		if sc.GetRunAsUser() == nil && sc.GetRunAsUsername() == "" && sc.GetRunAsGroup() != nil {
-			return fmt.Errorf("RunAsGroup should be specified only with RunAsUser or RunAsUsername")
-		}
-		groupstr := strconv.FormatInt(sc.GetRunAsGroup().GetValue(), 10)
-		if groupstr != "" {
-			containerUser = containerUser + ":" + groupstr
-		}
+	if sc == nil {
+		return nil
+	}
+	imageUser := ""
+	if imageConfig != nil {
+		imageUser = imageConfig.Config.User
+	}
+	containerUser, err := generateUserString(
+		sc.GetRunAsUsername(),
+		imageUser,
+		sc.GetRunAsUser(),
+		sc.GetRunAsGroup(),
+	)
+	if err != nil {
+		return err
+	}
 
-		logrus.Debugf("CONTAINER USER: %+v", containerUser)
+	logrus.Debugf("CONTAINER USER: %+v", containerUser)
 
-		// Add uid, gid and groups from user
-		uid, gid, addGroups, err1 := getUserInfo(rootfs, containerUser)
-		if err1 != nil {
-			return err1
-		}
+	// Add uid, gid and groups from user
+	uid, gid, addGroups, err := getUserInfo(rootfs, containerUser)
+	if err != nil {
+		return err
+	}
 
-		logrus.Debugf("UID: %v, GID: %v, Groups: %+v", uid, gid, addGroups)
-		specgen.SetProcessUID(uid)
-		specgen.SetProcessGID(gid)
-		for _, group := range addGroups {
-			specgen.AddProcessAdditionalGid(group)
-		}
+	logrus.Debugf("UID: %v, GID: %v, Groups: %+v", uid, gid, addGroups)
+	specgen.SetProcessUID(uid)
+	specgen.SetProcessGID(gid)
 
-		// Add groups from CRI
-		groups := sc.GetSupplementalGroups()
-		for _, group := range groups {
-			specgen.AddProcessAdditionalGid(uint32(group))
-		}
+	if sc.GetRunAsUsername() != "" {
+		containerUser = sc.GetRunAsUsername()
+	} else {
+		containerUser = strconv.FormatInt(sc.GetRunAsUser().GetValue(), 10)
+	}
+
+	_, _, addGroups2, err := getUserInfo(rootfs, containerUser)
```

Comment:
@cyphar We are having to call into GetExecUser twice to get the supplemental groups correctly. I don't like that we have to pass the files twice when we could return the groups as well in one go. This is for the case when the user string is set to 0:0. What do you think of adding a boolean so that we return groups as well?